src/bin/pgcopydb/ld_transform.c

```diff
@@ -2609,3 +2650,323 @@ LogicalMessageValueEq(LogicalMessageValue *a, LogicalMessageValue *b)
 	/* makes compiler happy */
 	return false;
 }
+
+
+/*
+ * Identifiers such as schema, table, column comes from various
+ * sources(e.g. wal2json, test_decoding and source catalog) and some of them
+ * already escapes identifiers and few don't.
+ * We need to check if the identifier is already quoted or not before
+ * escaping it.
+ * Whatever we are here is not a fool proof escaping mechanism, but a best
+ * effort to make sure that the identifiers are normalized by quoting them
+ * if it is not already quoted.
+ *
+ * Here is an example:
+ * foo -> "foo"
+ * "foo" -> "foo"
+ * foo"bar -> "foo"bar"
+ * "foo -> ""foo"
+ *
+ * The goal of this normalization is to make sure that the identifiers are
+ * comparable in the context of Hash Table.
+ */
+#define NORMALIZE_IDENTIFIER(dst, src, size) \
+	{ \
+		int len = strlen(src); \
+		if (src[0] == '"' && src[len - 1] == '"') \
+		{ \
+			strlcpy(dst, src, size); \
+		} \
+		else \
+		{ \
+			sformat(dst, size, "\"%s\"", src); \
+		} \
+	}
+
+/*
+ * findGeneratedColumns finds whether the given "nspname.relname" has any
+ * generated columns using the cache, if found it returns true and populates
+ * "columns" with the generated columns.
+ *
+ * NOTE: There is no error condition, if the cache is NULL, it means that we
+ * don't have any generated columns in the catalog.
+ */
+static bool
+findGeneratedColumns(GeneratedColumnsCache *cache,
+					 const char *nspname,
+					 const char *relname,
+					 GeneratedColumns **columns)
+{
+	/*
+	 * NULL cache means that we don't have any generated columns in the
+	 * catalog.
+	 */
+	if (cache == NULL)
+	{
+		return false;
+	}
+
+	GeneratedColumnsCache *item = NULL;
+
+	TableWithGeneratedColumns generatedColumn = { 0 };
+
+	NORMALIZE_IDENTIFIER(generatedColumn.nspname, nspname, PG_NAMEDATALEN);
+	NORMALIZE_IDENTIFIER(generatedColumn.relname, relname, PG_NAMEDATALEN);
+
+	HASH_FIND(hh, cache, &generatedColumn, sizeof(TableWithGeneratedColumns), item);
+
+	if (item == NULL)
+	{
+		return false;
+	}
+
+	if (item->columns == NULL)
+	{
+		log_error("BUG: Table \"%s.%s\" is in the cache but columns are NULL",
+				  nspname, relname);
+		return false;
+	}
+
+	*columns = item->columns;
+
+	log_debug("Table \"%s.%s\" has generated columns", nspname, relname);
```

Comment:
I suppose this would be `log_trace` if we think we want to keep it around in the merge?